src/hotspot/share/gc/g1/g1YoungGCPostEvacuateTasks.cpp

```diff
@@ -184,33 +184,30 @@ class G1FreeHumongousRegionClosure : public HeapRegionClosure {
     _freed_bytes(0) {
   }
 
-  virtual bool do_heap_region(HeapRegion* r) {
-    if (!r->is_starts_humongous()) {
+  bool do_heap_region_index(uint region_index) override {
+    if (!is_reclaimable(region_index)) {
       return false;
     }
 
-    uint region_idx = r->hrm_index();
-    if (!is_reclaimable(region_idx)) {
-      return false;
-    }
+    G1CollectedHeap* g1h = G1CollectedHeap::heap();
```

Comment:
Maybe look up the heap once at closure construction time, caching it in a data member?